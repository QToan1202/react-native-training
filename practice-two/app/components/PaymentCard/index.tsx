import { memo, useCallback, useMemo } from 'react'
import isEqual from 'react-fast-compare'
import { Stack, StackProps, YStack } from 'tamagui'

import Paragraph from '@components/Paragraph'
import { checkCardType, formatCardNumber } from '@utils'
import { ICardBase } from '@types'

import {
  StyledImage,
  StyledImageBackground,
  StyledImageBackgroundProps,
  StyledTitle,
} from './styles'

export type PaymentCardProps = StackProps &
  ICardBase & {
    isSelected?: boolean
    cardStyle?: StyledImageBackgroundProps['style']
    onCardSelected?: (data: ICardBase) => void
  }

const PaymentCard = ({
  name,
  number,
  cvc,
  expired,
  isSelected = false,
  onCardSelected,
  cardStyle,
  ...rest
}: PaymentCardProps) => {
  const cardBg = useMemo(() => {
    switch (checkCardType(number)) {
      case 'mastercard':
        return require('@assets/payment/mastercard.png')
      case 'visa':
        return require('@assets/payment/visa.png')
      default:
        return require('@assets/payment/normal.png')
    }
  }, [number])
  const handleSelectCard = useCallback(
    () => onCardSelected && onCardSelected({ name, number, cvc, expired }),
    [number, cvc, expired, name, onCardSelected]
  )

  return (
    <Stack
      alignSelf="flex-start"
      pressStyle={{ opacity: 0.8 }}
      onPress={handleSelectCard}
      {...rest}
    >
      <>
        <StyledImageBackground source={cardBg} style={cardStyle}>
          <YStack justifyContent="space-between" height="100%">
            <YStack>
              <StyledTitle content="holder name" />
              <Paragraph content={name} numberOfLines={1} />
            </YStack>

            <YStack>
              <StyledTitle content="card number" />
              <Paragraph
                fontSize="$2"
                width={167}
                numberOfLines={1}
                content={formatCardNumber(number)}
              />
            </YStack>

            <YStack>
              <StyledTitle content="exp. date" />
              <Paragraph content={expired} />
            </YStack>
          </YStack>

          <YStack>
            <StyledTitle content="cvc" textTransform="uppercase" />
            <Paragraph content={String(cvc)} textAlign="right" />
          </YStack>
        </StyledImageBackground>
        {isSelected && <StyledImage source={require('@assets/check.png')} />}
      </>
    </Stack>
  )
}

export default memo(PaymentCard, isEqual)
