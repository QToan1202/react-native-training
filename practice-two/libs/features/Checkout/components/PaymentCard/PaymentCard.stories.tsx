import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { PaymentCard } from '../index'

export default {
  title: 'components/Payment Card',
  component: PaymentCard,
} as ComponentMeta<typeof PaymentCard>

const Template: ComponentStory<typeof PaymentCard> = (args) => <PaymentCard {...args} />

export const Default = Template.bind({})
Default.args = {
  name: 'Tradly Team',
  cardNumber: '5627215898548869',
  cvc: '111',
  expires: '01/2019',
}

export const VisaCard = Template.bind({})
VisaCard.args = {
  name: 'Toan Nguyen',
  cardNumber: '4444444444444444',
  cvc: '312',
  expires: '12/2027',
}

export const SelectedCard = Template.bind({})
SelectedCard.args = {
  ...VisaCard.args,
  isSelected: true,
}
