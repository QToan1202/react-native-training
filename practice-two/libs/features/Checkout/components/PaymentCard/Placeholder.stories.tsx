import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { PaymentCardPlaceholder } from '../index'

export default {
  title: 'components/Payment Card Place Holder',
  component: PaymentCardPlaceholder,
} as ComponentMeta<typeof PaymentCardPlaceholder>

const Template: ComponentStory<typeof PaymentCardPlaceholder> = () => <PaymentCardPlaceholder />

export const Default = Template.bind({})
Default.args = {}
