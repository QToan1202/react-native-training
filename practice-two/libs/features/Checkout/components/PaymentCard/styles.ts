import { GetProps, styled } from 'tamagui'
import { Image, ImageBackground, StyleSheet } from 'react-native'

import Paragraph from '@components/Paragraph'
import { COLORS } from '@constants'

export const StyledImageBackground = styled(ImageBackground, {
  name: 'ImageBackground',
  source: { uri: '' },
  flexDirection: 'row',
  alignItems: 'flex-end',
  alignSelf: 'flex-start',
  paddingVertical: '$space.3',
  paddingHorizontal: 17,
  columnGap: 17,
  borderRadius: 9,
  overflow: 'hidden',
  maxHeight: 146,
  maxWidth: 246,
})

export type StyledImageBackgroundProps = GetProps<typeof StyledImageBackground>

export const StyledTitle = styled(Paragraph, {
  content: '',
  color: '$color.gray_200',
  fontSize: 13,
  textTransform: 'capitalize',
})

export type StyledTitleProps = GetProps<typeof StyledTitle>

export const StyledImage = styled(Image, {
  name: 'IconImage',
  source: { uri: '' },
  position: 'absolute',
  zIndex: '$zIndex.1',
  right: '$space.-0.5',
  bottom: '$space.-2.5',
})

export type StyledImageProps = GetProps<typeof StyledImage>

export const styles = StyleSheet.create({
  border: {
    paddingVertical: 36,
    paddingHorizontal: 38,
    alignItems: 'center',
    borderWidth: 2, // '$space.1'
    borderRadius: 10, // '$radius.5'
    borderColor: COLORS.GRAY_300, // '$color.gray_300'
    borderStyle: 'dashed',
  },
})
