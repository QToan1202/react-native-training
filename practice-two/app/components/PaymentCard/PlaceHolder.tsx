import { memo } from 'react'
import isEqual from 'react-fast-compare'
import { Image } from 'react-native'
import { View, YStackProps } from 'tamagui'

import Paragraph from '@components/Paragraph'
import { styles } from './styles'

export type PlaceHolderProps = YStackProps & {
  onTouchPlaceHolder?: () => void
}

const PlaceHolder = ({ onTouchPlaceHolder, ...rest }: PlaceHolderProps) => {
  return (
    <View style={styles.border} space={20} onPress={onTouchPlaceHolder} {...rest}>
      <Image source={require('@assets/payment/icon.png')} />
      <Paragraph
        color="$color.gray_300"
        textTransform="capitalize"
        lineHeight={28}
        letterSpacing={0.5}
        content="add payment method"
      />
    </View>
  )
}

export default memo(PlaceHolder, isEqual)
