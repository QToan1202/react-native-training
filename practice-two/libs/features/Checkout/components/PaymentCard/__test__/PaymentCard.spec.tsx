import { render, screen } from '@testing-library/react-native'

import PaymentCard from '../index'

describe('Testing Payment Card component', () => {
  beforeEach(() => {
    render(
      <PaymentCard
        name="card"
        cardNumber="4627215898548869"
        cvc={123}
        expires="01/2019"
        testID="payment-card"
      />
    )
  })

  it('Render Payment Card component success', () => {
    const component = render(
      <PaymentCard
        name="card"
        cardNumber="5627215898548869"
        cvc={123}
        expires="01/2019"
        isSelected
      />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('When render success component must exist on component tree', () => {
    const element = screen.getByTestId('payment-card')

    expect(element).toBeOnTheScreen()
  })

  it('Component must have source prop', () => {
    const element = screen.getByTestId('payment-card')

    expect(element).toHaveProp('source')
  })
})
