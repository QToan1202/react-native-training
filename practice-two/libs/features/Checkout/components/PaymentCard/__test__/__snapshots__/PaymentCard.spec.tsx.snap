// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Payment Card component Render Payment Card component success 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "flex-start",
      "opacity": 1,
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      [
        {
          "alignItems": "flex-end",
          "alignSelf": "flex-start",
          "borderRadius": 9,
          "columnGap": 17,
          "flexDirection": "row",
          "maxHeight": 146,
          "maxWidth": 246,
          "overflow": "hidden",
          "paddingHorizontal": 17,
          "paddingVertical": 13,
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/payment/mastercard.png",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "height": "100%",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              {
                "color": "#b7b7b7",
                "fontSize": 13,
                "textTransform": "capitalize",
              },
            ]
          }
        >
          holder name
        </Text>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              undefined,
            ]
          }
        >
          card
        </Text>
      </View>
      <View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              {
                "color": "#b7b7b7",
                "fontSize": 13,
                "textTransform": "capitalize",
              },
            ]
          }
        >
          card number
        </Text>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 16,
              },
              {
                "width": 167,
              },
            ]
          }
        >
          5627 2158 9854 8869 
        </Text>
      </View>
      <View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              {
                "color": "#b7b7b7",
                "fontSize": 13,
                "textTransform": "capitalize",
              },
            ]
          }
        >
          exp. date
        </Text>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              undefined,
            ]
          }
        >
          01/2019
        </Text>
      </View>
    </View>
    <View>
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            [
              {
                "color": "#b7b7b7",
                "fontSize": 13,
                "textTransform": "capitalize",
              },
              {
                "textTransform": "uppercase",
              },
            ],
          ]
        }
      >
        cvc
      </Text>
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "textAlign": "right",
            },
          ]
        }
      >
        123
      </Text>
    </View>
  </View>
  <Image
    source={
      {
        "testUri": "../../../assets/check.png",
      }
    }
    style={
      {
        "bottom": -10,
        "position": "absolute",
        "right": -1,
        "zIndex": 3,
      }
    }
  />
</View>
`;
